README.md
# Kourier

[![Docker Repository on Quay](https://quay.io/repository/3scale/kourier/status "Docker Repository on Quay")](https://quay.io/repository/3scale/kourier)

Kourier is an Ingress for [Knative Serving](https://knative.dev/). Kourier is a
lightweight alternative for the Istio ingress as its deployment consists only of
an Envoy proxy and a control plane for it.

Kourier is passing the knative serving e2e and conformance tests:
[Kourier Testgrid](https://testgrid.knative.dev/serving#kourier-stable).

- [**Getting started**](#getting-started)
- [**Features**](#features)
- [**Development**](#development)
- [**License**](#license)

## Getting started

- Install Knative Serving, ideally without Istio by using the
  "serving-core.yaml":

```bash
kubectl apply -f https://github.com/knative/serving/releases/download/v0.12.1/serving-core.yaml
```

- Then install Kourier:

```bash
kubectl apply -f deploy/kourier-knative.yaml
```

- Configure Knative Serving to use the proper "ingress.class":

```bash
kubectl patch configmap/config-network \
  -n knative-serving \
  --type merge \
  -p '{"data":{"ingress.class":"kourier.ingress.networking.knative.dev"}}'
```

- (OPTIONAL) Set your desired domain (replace 127.0.0.1.nip.io to your prefered
  domain):

```bash
 kubectl patch configmap/config-domain \
  -n knative-serving \
  --type merge \
  -p '{"data":{"127.0.0.1.nip.io":""}}'
```

- (OPTIONAL) Deploy a sample hello world app:

```bash
kubectl apply -f ./samples/helloworld-go.yaml
```

- (OPTIONAL) For testing purposes, you can use port-forwarding to make requests
  to Kourier from your machine:

```bash
kubectl port-forward --namespace kourier-system $(kubectl get pod -n kourier-system -l "app=3scale-kourier-gateway" --output=jsonpath="{.items[0].metadata.name}") 8080:8080 19000:19000 8443:8443

curl -v -H "Host: helloworld-go.default.127.0.0.1.nip.io" http://localhost:8080
```

## Features

- Traffic splitting between Knative revisions.
- Automatic update of endpoints as they are scaled.
- Support for gRPC services.
- Timeouts and retries.
- TLS
- External Authorization support.

## Setup TLS certificate

Create a secret containing your TLS certificate and Private key:

```
kubectl create secret tls ${CERT_NAME} --key ${KEY_FILE} --cert ${CERT_FILE}
```

Add the following env vars to 3scale-Kourier in the "kourier" container :

```
CERTS_SECRET_NAMESPACE: ${NAMESPACES_WHERE_THE_SECRET_HAS_BEEN_CREATED}
CERTS_SECRET_NAME: ${CERT_NAME}
```

## External Authorization Configuration

If you want to enable the external authorization support you can set these ENV
vars in the `3scale-kourier-control` deployment:

- `KOURIER_EXTAUTHZ_HOST*`: The external authorization service and port,
  my-auth:2222
- `KOURIER_EXTAUTHZ_FAILUREMODEALLOW*`: Allow traffic to go through if the ext
  auth service is down. Accepts true/false
- `KOURIER_EXTAUTHZ_MAXREQUESTBYTES`: Max request bytes, if not set, defaults to
  8192 Bytes. More info
  [Envoy Docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto.html?highlight=max_request_bytes#extensions-filters-http-ext-authz-v3-buffersettings)
- `KOURIER_EXTAUTHZ_TIMEOUT`: Max time in ms to wait for the ext authz service.
  Defaults to 2s.

`*` Required

## Development

- Run the test suite:

```bash
make test
```

- Run only the unit or the integration tests:

```bash
make test-unit
make test-integration
```

- Set up a local environment with Knative running on top of
  [k3s](https://k3s.io/):

```bash
make local-setup
```

- Run `make help` for the complete list of make targets available.

## License

[Apache 2.0 License](LICENSE)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.12] - 2020-03-10

### Fixed

- Overall reliability improvements, now kourier resyncs failed ingresses every
  30s.

## [0.3.11] - 2020-03-06

### Fixed

- Now when Envoy rejects a configuration, Kourier retries it.

### Changed

- Now the gateway pod uses `docker.io/maistra/proxyv2-ubi8:1.0.8`.

## [0.3.10] - 2020-02-28

### Added

- Envoy External Authz support. Now you can define an external point for
  authorizing traffic.

### Fixed

- Kourier control will not try to probe non ready gateways.
- If there's no gateway available, the event will be retried.

## [0.3.9] - 2020-02-20

### Added

- Added more debug logging.

### Fixed

- Fixed a race condition when the endpoints aren't created fast enough.
- Envoy configuration is now refreshed when a cluster expires from the config
  cache.

## [0.3.8] - 2020-01-24

### Fixed

- Fixed a situation where routes were created empty. Causing test flakiness and
  unexpected user errors.

## [0.3.7] - 2020-01-22

### Fixed

- Improved handling of service updates by using RDS. Some 5xx errors that
  appeared when updating a service should no longer occur.
- Fixed a bug when using auto TLS. In some cases, Kourier failed to expose some
  services.

## [0.3.6] - 2020-01-14

### Fixed

- Expose HTTP and HTTPS in the same service.

## [0.3.5] - 2020-01-13

### Added

- Prometheus stats endpoint exposed.
- Support for SNI-based routing.

### Changed

- Simplified deployment templates.
- Changed the Kourier namespace to `kourier-system`.

### Fixed

- Ingresses are now correctly reconciled when deleted.
- As defined in Knative Serving, headers are now replaced instead of appended
  when proxying the request.

## [0.3.4] - 2019-12-11

### Security

- Updated Kourier Gateway to use envoy version 1.12.2

## [0.3.3] - 2019-12-05

### Fixed

- Bug that caused some Envoy clusters to be deleted when they were still
  referenced by a route.

## [0.3.2] - 2019-12-04

### Changed

- Checking whether an ingress should be marked as ready is no longer done
  online. It's done in separate go routines.

### Fixed

- "concurrent map writes" errors caused by an incorrect usage of locks.

## [0.3.1] - 2019-12-03

### Changed

- Instead of refreshing the whole Envoy config, now Kourier updates only the
  parts that affect the modified ingress or endpoint.

## [0.3.0] - 2019-11-29

### Changed

- Adapted the codebase to use Knative's controllers and reconcilers.

## [0.2.6] - 2019-11-14

### Fixed

- Kourier now only routes the public endpoints object of a revision.

## [0.2.5] - 2019-11-13

### Changed

- Added readiness probe to the kourier gateway pod.

## [0.2.4] - 2019-11-13

### Fixed

- Now a knative Ingress is not marked as ready until all the kourier gateways
  are in sync with the latest configuration version.

## [0.2.3] - 2019-11-11

### Removed

- Dropped support for ClusterIngress CRD, which means dropping support for
  Knative Serving < 0.9.

## [0.2.2] - 2019-11-06

### Fixed

- Previous "missing Route" fix was not covering all the cases. Now it's fixed
  with the implementation of cache for clusters, details can be found in the
  source code.

## [0.2.1] - 2019-11-04

### Changed

- Updated envoy go control plane dependency to v0.9.0
- Get the cluster local domain automatically from the "/etc/resolv.conf" file
- Replaced deprecated instructions from envoy bootstrap config.

### Fixed

- Fix for a "missing Route" issue when revisions where replaced/modified too
  quickly.

## [0.2.0] - 2019-10-25

### Added

- New kourier-gateway docker image

### Changed

- Splitted the kourier POD into two new PODS, kourier-control and
  kourier-gateway.
- Kourier filters knative serving ingress.class, looks for:
  'kourier.ingress.networking.knative.dev'.

## [0.1.0] - 2019-10-23

First release.

[0.3.12]: https://github.com/knative/net-kourier/compare/v0.3.11...v0.3.12
[0.3.11]: https://github.com/knative/net-kourier/compare/v0.3.10...v0.3.11
[0.3.10]: https://github.com/knative/net-kourier/compare/v0.3.9...v0.3.10
[0.3.9]: https://github.com/knative/net-kourier/compare/v0.3.8...v0.3.9
[0.3.8]: https://github.com/knative/net-kourier/compare/v0.3.7...v0.3.8
[0.3.7]: https://github.com/knative/net-kourier/compare/v0.3.6...v0.3.7
[0.3.6]: https://github.com/knative/net-kourier/compare/v0.3.5...v0.3.6
[0.3.5]: https://github.com/knative/net-kourier/compare/v0.3.4...v0.3.5
[0.3.4]: https://github.com/knative/net-kourier/compare/v0.3.3...v0.3.4
[0.3.3]: https://github.com/knative/net-kourier/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/knative/net-kourier/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/knative/net-kourier/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/knative/net-kourier/compare/v0.2.6...v0.3.0
[0.2.6]: https://github.com/knative/net-kourier/compare/v0.2.5...v0.2.6
[0.2.5]: https://github.com/knative/net-kourier/compare/v0.2.4...v0.2.5
[0.2.4]: https://github.com/knative/net-kourier/compare/v0.2.3...v0.2.4
[0.2.3]: https://github.com/knative/net-kourier/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/knative/net-kourier/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/knative/net-kourier/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/knative/net-kourier/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/knative/net-kourier/releases/tag/v0.1.0


utils/setup.sh
#!/bin/bash
set -euo pipefail
IFS=$'\n\t'

KOURIER_NAMESPACE=kourier-system
KNATIVE_NAMESPACE=knative-serving

if ! command -v k3d >/dev/null; then
  echo "k3d binary not in path, install with: curl -s https://raw.githubusercontent.com/rancher/k3d/master/install.sh | bash"
  exit 1
fi

tag="test_$(git rev-parse --abbrev-ref HEAD)"

# In CircleCI, PR branches that come from forks have the format "pull/n", where
# n is the PR number. "/" is not accepted in docker tags, so we need to replace
# it.
tag=$(echo "$tag" | tr / -)

k3d d --name=kourier-integration || true

k3d c --name kourier-integration --server-arg "--no-deploy=traefik"
sleep 60
export KUBECONFIG="$(k3d get-kubeconfig --name='kourier-integration')"

# Builds and imports the kourier and gateway images from docker into the k8s cluster
docker build -t 3scale-kourier:"$tag" ./
docker build -f ./utils/extauthz_test_image/Dockerfile -t test_externalauthz:test ./utils/extauthz_test_image/
k3d import-images 3scale-kourier:"$tag" --name='kourier-integration'
k3d import-images test_externalauthz:test --name='kourier-integration'

KNATIVE_VERSION=v0.13.2
# Deploys kourier and patches it.
kubectl apply -f https://github.com/knative/serving/releases/download/${KNATIVE_VERSION}/serving-crds.yaml
kubectl apply -f https://github.com/knative/serving/releases/download/${KNATIVE_VERSION}/serving-core.yaml
kubectl apply -f deploy/kourier-knative.yaml
kubectl patch deployment 3scale-kourier-control -n ${KOURIER_NAMESPACE} --patch "{\"spec\": {\"template\": {\"spec\": {\"containers\": [{\"name\": \"kourier-control\",\"image\": \"3scale-kourier:$tag\",\"imagePullPolicy\": \"IfNotPresent\"}]}}}}"
kubectl patch configmap/config-domain -n ${KNATIVE_NAMESPACE} --type merge -p '{"data":{"127.0.0.1.nip.io":""}}'
kubectl patch configmap/config-network -n ${KNATIVE_NAMESPACE} --type merge -p '{"data":{"ingress.class":"kourier.ingress.networking.knative.dev"}}'

retries=0
while [[ $(kubectl get pods -n ${KOURIER_NAMESPACE} -l app=3scale-kourier-control -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}') != "True" ]]; do
  echo "Waiting for kourier control pod to be ready "
  sleep 10
  if [ $retries -ge 7 ]; then
    echo "timedout waiting for kourier control pod"
    exit 1
  fi
  retries=$((retries + 1))
done

retries=0
while [[ $(kubectl get pods -n ${KOURIER_NAMESPACE} -l app=3scale-kourier-gateway -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}') != "True" ]]; do
  echo "Waiting for kourier gateway pod to be ready "
  sleep 10
  if [ $retries -ge 7 ]; then
    echo "timedout waiting for kourier gateway pod"
    exit 1
  fi
  retries=$((retries + 1))
done

# shellcheck disable=SC2046
kubectl port-forward --namespace ${KOURIER_NAMESPACE} "$(kubectl get pod -n ${KOURIER_NAMESPACE} -l "app=3scale-kourier-gateway" --output=jsonpath="{.items[0].metadata.name}")" 8080:8080 8081:8081 19000:19000 8443:8443 &>/dev/null &
